### PR TITLE
add DART: Dropouts meet Multiple Additive Regression Trees

### PR DESCRIPTION
**Pull request by wxchan**
Related paper: [DART: Dropouts meet Multiple Additive Regression Trees](https://arxiv.org/abs/1505.01866)

Configuration:
boosting_type = dart
drop_rate: Rate for selecting dropping trees (default 0.01, type = double, 0<=drop_rate<=1, 0 means randomly select one tree) 
dropping_seed: Random seed for selecting dropping trees (default 4, type = int)

Performance:
Speed: about 1.5-2 times slower
Accuracy:
Higgs: auc = 0.850505, drop_rate = 0.01
Yahoo LTR: ndcg@1,3,5,10: 0.72901 0.73677 0.75545 0.79537, drop_rate = 0.03
MS LTR: ndcg@1,3,5,10: 0.522249 0.506179 0.510583 0.528597, drop_rate = 0.02

### CODE REVIEW

File: src/boosting/dart.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+     * first, shrink the dropping trees to m - 1 = -n
+     * update the score, the dropping trees will weight 1 + (m - 1) = m
+     * then, shrink the dropping trees to m / (m - 1) = -k
+     */
+    double k = static_cast<double>(drop_index.size());
+    double new_tree_shrinkage_rate = 1.0 / (1.0 + k);
+    for (int i: drop_index){
+        models_[i]->Shrinkage(-new_tree_shrinkage_rate);
+        UpdateScore(models_[i], curr_class);
+        models_[i]->Shrinkage(-k);
```
Comment (guolinke) [inline src/boosting/dart.cpp]: Is this correct? Can we predict the same score with these saved model?

Comment (wxchan) [inline src/boosting/dart.cpp]: @guolinke not sure what you mean, I change the code [here](https://github.com/Microsoft/LightGBM/pull/67/files#diff-0f191b18c888fb11d844c8a5e8cb01adR251), the DART model only saves to file once.

File: src/boosting/dart.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+     * then, shrink the dropping trees to m / (m - 1) = -k
+     */
+    double k = static_cast<double>(drop_index.size());
+    double new_tree_shrinkage_rate = 1.0 / (1.0 + k);
+    for (int i: drop_index){
+      for (int curr_class = 0; curr_class < num_class_; ++curr_class) {
+        int curr_tree = i * num_class_ + curr_class;
+        models_[curr_tree]->Shrinkage(-new_tree_shrinkage_rate);
+        UpdateScore(models_[curr_tree], curr_class);
+        models_[curr_tree]->Shrinkage(-k);
```
Comment (guolinke) [inline src/boosting/dart.cpp]: According the algorithm 1 in Paper. I think here is:
```
// remove the score of curr_tree
models_[curr_tree]->Shrinkage(-1.0);
UpdateScore(models_[curr_tree], curr_class);
// re-factor by k / ( k + 1.0)
models_[curr_tree]->Shrinkage(- k / (k + 1.0));
```
Did I miss something?

File: src/boosting/dart.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  return 1.0 / (1.0 + drop_index_.size());
+}
+
+void DART::Normailize(double shrinkage_rate) {
+  double k = static_cast<double>(drop_index_.size());
+  for (int i: drop_index_){
+    for (int curr_class = 0; curr_class < num_class_; ++curr_class) {
+      int curr_tree = i * num_class_ + curr_class;
+      models_[curr_tree]->Shrinkage(-k * shrinkage_rate);
+      UpdateScore(models_[curr_tree], curr_class);
```
Comment (guolinke) [inline src/boosting/dart.cpp]: Do we need update score at here? 
I am not sure about this, and the paper doesn't give the clear explanation. 
Did you have any test?

Comment (guolinke) [inline src/boosting/dart.cpp]: It seems we need to update score here, to get the real prediction score from ensemble models.

Comment (wxchan) [inline src/boosting/dart.cpp]: @guolinke I have a little misunderstanding before. I think the logic is:

1. drop trees: 
models_[curr_tree]->Shrinkage(-1.0);
UpdateScore(models_[curr_tree], curr_class);

2. calculate gradient and hessian

3. train new trees and normalize new trees:
new_tree->Shrinkage(1 / (1 + k));
UpdateScore(new_tree, curr_class);

4. normalize dropped trees:
models_[curr_tree]->Shrinkage(- k / (k + 1.0));
UpdateScore(models_[curr_tree], curr_class);

5. eval metrics

The order of step 1 and 2 is not correct now. I am doing some test. I will tell you when I finish.

File: src/boosting/dart.cpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+bool DART::TrainOneIter(const score_t* gradient, const score_t* hessian, bool is_eval) {
+  // boosting first
+  if (gradient == nullptr || hessian == nullptr) {
+    Boosting();
+    gradient = gradients_;
+    hessian = hessians_;
+  }
+  
+  // drop trees
+  double shrinkage_rate = DroppingTrees();
```
Comment (guolinke) [inline src/boosting/dart.cpp]: I think this is not correct. We need to drop tree before ``Boosting``. 
The purpose of drop-out is for the gradients computation, which is done in ``Boosting``.

File: src/boosting/dart.cpp (hunk 3 above)
Comment (guolinke) [inline src/boosting/dart.cpp]: OK, BTW, It seems we cannot support self-defined gradients/hessians directly in DART.

File: src/boosting/dart.cpp (hunk 4 above)
Comment (wxchan) [inline src/boosting/dart.cpp]: @guolinke yes, you are right. I already changed this and am doing the test. See my comment above.

File: src/boosting/dart.cpp (hunk 3 above)
Comment (wxchan) [inline src/boosting/dart.cpp]: Is it possible we can calculate delta gradients/hessians before and after dropping the trees and add them into self-defined gradients/hessians?

Comment (guolinke) [inline src/boosting/dart.cpp]: I don't think so. The self-defined gradients/hessians is calculated by unknown objective function. I think we should disable drop-out when passed gradients/hessians is nullptr.

**Comment by guolinke** [comment]
@wxchan  can you tried merge this to dev branch as well?

**Comment by wxchan** [comment]
@guolinke How about I do it after you merge dev into master? I can take a few more days to investigate if there is anything wrong.

But I think current logic is right. 

New test performance:
Speed: about 1.5-2 times slower
Accuracy:
Higgs: auc = 0.850505 (> 0.845123), drop_rate = 0.01
Yahoo LTR: ndcg@1,3,5,10: 0.72901 0.73677 0.75545 0.79537, drop_rate = 0.03; 0.02-0.03 worse
MS LTR: ndcg@1,3,5,10: 0.522249 0.506179 0.510583 0.528597, drop_rate = 0.02, a little improvement, not impressive

Do you think it's better to put Class Dart into gbdt.cpp/h, because it's kinda a regularization method for gbdt?

**Comment by guolinke** [comment]
@wxchan I think it is ok to put DART to gbdt.h/cpp. However, let us keep using independent file now, for i have changed some interface in gbdt.h/cpp, which may cause conflict.

File: src/application/application.cpp [hunk 5]
```diff
@@ -248,14 +248,17 @@ void Application::Train() {
   int total_iter = config_.boosting_config->num_iterations;
   bool is_finished = false;
   bool need_eval = true;
+  bool save_model_during_training = (config_.boosting_type != BoostingType::kDART);
```
Comment (guolinke) [inline src/application/application.cpp]: move this inside boosting->SaveModelToFile ?

File: src/boosting/dart.cpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+     const std::vector<const Metric*>& training_metrics) {
+  GBDT::Init(config, train_data, object_function, training_metrics);
+  gbdt_config_ = dynamic_cast<const GBDTConfig*>(config);
+  drop_rate_ = gbdt_config_->drop_rate;
+  random_for_drop_ = Random(gbdt_config_->dropping_seed);
+}
+
+bool DART::TrainOneIter(const score_t* gradient, const score_t* hessian, bool is_eval) {
+  // drop trees
+  double shrinkage_rate = DroppingTrees();
```
Comment (guolinke) [inline src/boosting/dart.cpp]: I think the dropout is need only when we need ``Boosting()`` . 
move this before Line:40?

Comment (wxchan) [inline src/boosting/dart.cpp]: OK. But in this case, there is no need to use dart rather than gbdt. I forgot to say this: I think it's still doable to use self-defined gradients if the case is self-defined objective function. Because user still needs 'scores' (or called current prediction) to calculate the gradients. We can drop the trees, update the scores, pass back the scores to somewhere and calculate gradients, and do the remaining steps. I am not sure how you implement self-defined objective function now, so I will leave this now and change it with other together.

Comment (guolinke) [inline src/boosting/dart.cpp]: @wxchan https://github.com/Microsoft/LightGBM/blob/dev/src/boosting/gbdt.cpp#L351
You can add the drop-out logic into this GetTrainingScore.

**Comment by guolinke** [comment]
@wxchan , The c_api is finished now. Can you continue to DART ?  Thanks

**Comment by wxchan** [comment]
@guolinke I fixed conflicts. I think GetTrainingScore() probably should remain const, so I make train_score_updater mutable, is there a better way to solve this?

**Comment by guolinke** [comment]
@wxchan I think it is better to remove ``const``

**Comment by wxchan** [comment]
@guolinke then `const` in [this line](https://github.com/Microsoft/LightGBM/blob/master/src/c_api.cpp#L130) will be removed too because it's used in [LGBM_BoosterGetScore](https://github.com/Microsoft/LightGBM/blob/master/src/c_api.cpp#L481-L483). Is it ok? Or is there any step before TrainOneIter to check whether user defined an objective function?

**Comment by guolinke** [comment]
@wxchan I have updated the code in master branch.

**Comment by wxchan** [comment]
@guolinke `const` and `mutable` are removed. All done.
